### PR TITLE
Solve performance regresion

### PR DESCRIPTION
**Pull request by lbenet**
See #102 for details of the problem.

### CODE REVIEW

**Comment by lbenet** [comment]
@dpsanders Can you review this PR? It seems that #102 is solved with the last commit.

**Comment by dpsanders** [comment]
Yes, I should be able to review it tonight or tomorrow.

**Comment by lbenet** [comment]
Thanks, that would be great!

File: perf/fateman.jl [hunk 1]
```diff
@@ -10,9 +10,9 @@ function fateman1(degree::Int)
     s = TaylorN( [oneH, HomogeneousPolynomial([one(T),one(T),one(T),one(T)],1)], degree )
     s = s^degree
     # s is converted to order 2*ndeg
-    s = TaylorN(s, 2*degree)
+    s = TaylorN(s.coeffs, 2*degree)
```
Comment (dpsanders) [inline perf/fateman.jl]: Why is this change to `s.coeffs` necessary?

File: perf/fateman.jl [hunk 2]
```diff
@@ -10,9 +10,9 @@ function fateman1(degree::Int)
     s = TaylorN( [oneH, HomogeneousPolynomial([one(T),one(T),one(T),one(T)],1)], degree )
```
Comment (dpsanders) [inline perf/fateman.jl]: `s = 1 + w + x + y + z`?

Comment (dpsanders) [inline perf/fateman.jl]: Why don't you fix the degree here to be `2*degree` already?

File: perf/fateman.jl [hunk 3]
```diff
@@ -22,7 +22,7 @@ function fateman2(degree::Int)
...
     s = s^degree
     # s is converted to order 2*ndeg
-    s = TaylorN(s, 2*degree)
+    s = TaylorN(s.coeffs, 2*degree)
```
Comment (dpsanders) [inline perf/fateman.jl]: Same comments here.

File: src/arithmetic.jl [hunk 4]
```diff
@@ -15,18 +15,10 @@ for T in (:Taylor1, :TaylorN)
...
-            lb = b.order+1
-            if a.order == b.order
-                return all( a.coeffs .== b.coeffs )
-            elseif a.order < b.order
-                res1 = all( a[1:la] .== b[1:la] )
-                res2 = iszero(b[la+1:lb])
-            else a.order > b.order
-                res1 = all( a[1:lb] .== b[1:lb] )
-                res2 = iszero(a[lb+1:la])
+            if a.order != b.order
```
Comment (dpsanders) [inline src/arithmetic.jl]: I suggest putting the `a.order != b.order` test inside `fixorder`, and just doing `a,b = fixorder(a, b)` here.

File: src/arithmetic.jl [hunk 5]
```diff
@@ -40,17 +32,11 @@ end
 
 ## zero and one ##
 for T in (:Taylor1, :TaylorN), f in (:zero, :one)
-    @eval begin
-        ($f)(a::$T) = $T([($f)(a[1])], a.order)
-
-        ($f)(a::$T, order) = $T([($f)(a[1])], order)
-    end
+    @eval ($f)(a::$T) = $T(($f)(a[1]), a.order)
```
Comment (dpsanders) [inline src/arithmetic.jl]: (In the future) we should think if `TaylorN` can literally be written in terms of `Taylor1` to avoid all this duplication of similar methods.

File: src/arithmetic.jl [hunk 6]
```diff
@@ -96,19 +80,12 @@ for (f, fc) in ((:+, :(add!)), (:-, :(subst!)))
...
-                    @inbounds @__dot__ v[1:la] = $f(a[1:la], b[1:la])
-                    @inbounds @__dot__ v[la+1:lb] = $f(b[la+1:lb])
-                else
-                    @inbounds @__dot__ v[1:lb] = $f(a[1:lb], b[1:lb])
-                    @inbounds @__dot__ v[lb+1:la] = a[lb+1:la]
+                if a.order != b.order
+                    a, b = fixorder(a, b)
                 end
-                return $T(v)
+                v = similar(a.coeffs)
```
Comment (dpsanders) [inline src/arithmetic.jl]: Why do you need this? Can't you just do `v = $f(a.coeffs, b.coeffs)` straight away?

File: src/arithmetic.jl [hunk 7]
```diff
@@ -280,10 +256,11 @@ function *{T<:NumberNotSeriesN}(a::HomogeneousPolynomial{T}, b::HomogeneousPolyn
...
 
-# Homogeneous coefficient for the multiplication
+# Internal multiplication functions
 for T in (:Taylor1, :TaylorN)
-    @eval function mul!(c::$T, a::$T, b::$T, k::Int)
+    @eval @inline function mul!(c::$T, a::$T, b::$T, k::Int)
 
+        # c[k+1] = zero( a[k+1] )
         @inbounds for i = 0:k
             if $T == Taylor1
```
Comment (dpsanders) [inline src/arithmetic.jl]: I suggest defining `mul!` for scalars instead to remove this special case.

File: src/arithmetic.jl [hunk 8]
```diff
@@ -317,7 +294,7 @@ c_k = \sum_{j=0}^k a_j b_{k-j}.
 Return `c = a*b` with no allocation; all arguments are `HomogeneousPolynomial`.
 
 """
-function mul!(c::HomogeneousPolynomial, a::HomogeneousPolynomial,
+@inline function mul!(c::HomogeneousPolynomial, a::HomogeneousPolynomial,
         b::HomogeneousPolynomial)
 
     (iszero(b) || iszero(a)) && return nothing
```
Comment (dpsanders) [inline src/arithmetic.jl]: You can just use `return`.

File: src/auxiliary.jl [hunk 9]
```diff
@@ -18,7 +18,8 @@ function resize_coeffs1!{T<:Number}(coeffs::Array{T,1}, order::Int)
     lencoef = length(coeffs)
     order ≤ lencoef-1 && return nothing
     resize!(coeffs, order+1)
-    coeffs[lencoef+1:end] .= zero(coeffs[1])
+    z = zero(coeffs[1])
+    @__dot__ coeffs[lencoef+1:order+1] = z
```
Comment (dpsanders) [inline src/auxiliary.jl]: This change is not needed I think.

Comment (dpsanders) [inline src/auxiliary.jl]: It does the same as the simpler version. Or even do `coeffs[lencoef+1:end] .= 0`?

File: src/auxiliary.jl [hunk 10]
```diff
@@ -35,7 +36,8 @@ function resize_coeffsHP!{T<:Number}(coeffs::Array{T,1}, order::Int)
     @assert order ≤ get_order() && lencoef ≤ num_coeffs
     num_coeffs == lencoef && return nothing
     resize!(coeffs, num_coeffs)
-    coeffs[lencoef+1:end] .= zero(coeffs[1])
+    z = zero(coeffs[1])
+    @__dot__ coeffs[lencoef+1:num_coeffs] = z
```
Comment (dpsanders) [inline src/auxiliary.jl]: Same comment.

File: src/auxiliary.jl [hunk 11]
```diff
@@ -71,10 +73,7 @@ Return the coefficient of order `n::Int` of a `a::Taylor1` polynomial.
 get_coeff(a::Taylor1, n::Int) = (@assert 0 ≤ n ≤ a.order+1;
     return a[n+1])
 
-function getindex(a::Taylor1, n::Int)
-    (1 ≤ n ≤ length(a.coeffs)) && return a.coeffs[n]
-    return zero(a.coeffs[1])
-end
+getindex(a::Taylor1, n::Int) = a.coeffs[n]
```
Comment (dpsanders) [inline src/auxiliary.jl]: No bounds check required?

File: src/auxiliary.jl [hunk 12]
```diff
@@ -167,17 +162,22 @@ for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
...
-For `a::Taylor1` returns `a.order` while for `a::TaylorN` returns
-`get_order()`.
-"""
-max_order(a::Taylor1) = a.order
+## fixorder ##
+for T in (:Taylor1, :TaylorN)
+    @eval begin
+        function fixorder(a::$T, b::$T)
+            a.order == b.order && return a, b
+            a.order < b.order && return $T(copy(a.coeffs), b.order),
```
Comment (dpsanders) [inline src/auxiliary.jl]: No `copy` required here?

File: src/auxiliary.jl [hunk 13]
```diff
@@ -167,17 +162,22 @@ for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
...
+                $T(copy(b.coeffs), b.order)
+            return $T(copy(a.coeffs), a.order), $T(copy(b.coeffs), a.order)
```
Comment (dpsanders) [inline src/auxiliary.jl]: Or here?

File: src/constructors.jl [hunk 14]
```diff
@@ -31,16 +31,19 @@ immutable Taylor1{T<:Number} <: AbstractSeries{T}
     ## Inner constructor ##
     function (::Type{Taylor1{T}}){T}(coeffs::Array{T,1}, order::Int)
         resize_coeffs1!(coeffs, order)
-        return new{T}(coeffs, length(coeffs)-1)
+        return new{T}(coeffs, order)
```
Comment (dpsanders) [inline src/constructors.jl]: Do you need `@compat` here?

File: src/power.jl [hunk 15]
```diff
@@ -40,6 +40,34 @@ for T in (:Taylor1, :TaylorN)
 end
 
 
+# power_by_squaring; slightly modified from base/intfuncs.jl
```
Comment (dpsanders) [inline src/power.jl]: What is the modification? Why is it required?

File: src/power.jl [hunk 16]
```diff
@@ -184,31 +211,33 @@ c_k & = & 2 \\sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2,
...
+        @inline function sqr!(c::$T, a::$T, k::Int)
+            if k == 0
+                @inbounds c[1] = constant_term(a)^2
+                return nothing
+            end
+
+            kodd = k%2
+            kend = div(k - 2 + kodd, 2)
+            @inbounds for i = 0:kend
+                if $T == Taylor1
```
Comment (dpsanders) [inline src/power.jl]: Use `mul!`

File: src/power.jl [hunk 17]
```diff
@@ -184,31 +211,33 @@ c_k & = & 2 \\sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2,
...
+            end
+            @inbounds c[k+1] = 2 * c[k+1]
+            kodd == 1 && return nothing
 
-        kodd = k%2
-        kend = div(k - 2 + kodd, 2)
-        @inbounds for i = 0:kend
             if $T == Taylor1
-                c[k+1] += a[i+1] * a[k-i+1]
+                @inbounds c[k+1] += a[div(k,2)+1]^2
```
Comment (dpsanders) [inline src/power.jl]: Define `sqr!` for scalars and remove the special case.

Comment (dpsanders) [inline src/power.jl]: (These tiny functions will be inlined.)

File: test/fateman40.jl [hunk 18]
```diff
@@ -15,7 +15,7 @@ using Base.Test
             [oneH, HomogeneousPolynomial([one(T),one(T),one(T),one(T)],1)], degree)
         s = s^degree
         # s is converted to order 2*ndeg
-        s = TaylorN(s, 2*degree)
+        s = TaylorN(s.coeffs, 2*degree)
```
Comment (dpsanders) [inline test/fateman40.jl]: Again not sure why this change is necessary.

File: perf/fateman.jl (hunk 2 above)
Comment (lbenet) [inline perf/fateman.jl]: The correct answer is "for historical reasons", which somehow do not apply any longer. As it is now, when you compute `s^degree` it only calculates up to `order=degree`, instead of putting lots of zeros (which are homogeneous polynomials of the correct order). I checked the impact in performance of putting `2*degree` there and, with respect to time there is none, but memory increases.

If you want, I can change it, but I don't think it is necessary for this PR.

File: perf/fateman.jl (hunk 1 above)
Comment (lbenet) [inline perf/fateman.jl]: The reason is related to a method I deleted of `TaylorN` (as well as the corresponding for `Taylor1` and `HomogeneousPolynomial`). The deleted method (in constructors.jl) was `TaylorN{T<:Number}(x::TaylorN{T}, order::Int) = TaylorN{T}(x.coeffs, order)`. Those methods have side effects, which may yield some unexpected behaviour...

File: src/arithmetic.jl (hunk 4 above)
Comment (lbenet) [inline src/arithmetic.jl]: I tested that option and retested it now again including `@inline` for `fixorder`, and the benchmarks are slightly worst with respect to time, due to more memory use. I am leaving this as it is.

File: src/arithmetic.jl (hunk 5 above)
Comment (lbenet) [inline src/arithmetic.jl]: I fully agree. I have done some work along this, but it is largely unfinished. Actually, after reading different sources, it seems to me that it is better (in terms of memory) to have `Taylor1{Taylor1{T}}`, which effectively is a `TaylorN{T}` of two variables.

File: src/arithmetic.jl (hunk 6 above)
Comment (lbenet) [inline src/arithmetic.jl]: You certainly can simply do `v = $f(a.coeffs, b.coeffs)`, but as far as I understand, using `@__dot__` avoids creating some temporary arrays and copying the results, which is therefore faster. For the type of benchmarks that Jorge was checking in #102, without `@__dot__` in the addition functions produces ~30% more memory usage, and ~10% slower code.

Comment (dpsanders) [inline src/arithmetic.jl]: In general, I agree. In this particular case, though, I don't see where the temporary will be.

File: src/arithmetic.jl (hunk 7 above)
Comment (lbenet) [inline src/arithmetic.jl]: We can certainly define a `mul` like method that returns `c[k+1] += a[i+1] * b[k-i+1]`. But my guess is that it will get inlined and produce the same code. Surely this will avoid the ugly code separating the `Taylor1` and `TaylorN` cases, but this happens at parse time, so I don't think we should care too much on this.

File: src/arithmetic.jl (hunk 6 above)
Comment (lbenet) [inline src/arithmetic.jl]: After reading your comment I also thought that this should produce no temporary array, but after checking it, it turn out that it does!

File: src/auxiliary.jl (hunk 9 above)
Comment (lbenet) [inline src/auxiliary.jl]: It is not really needed, you are right, but it yields a tiny performance improvement (at a very low cost in memory). The change you propose (`coeffs[lencoef+1:end] .= 0`) could yield the same or even better performance; yet, for `Taylor1{Taylor1{T}}` cases, and other mixtures, using `0` instead of `zero(coeffs[1])` produces slightly different things with respect to the order of the first coefficient. While this in general should not be important, I can think of cases where it may have consequences in the sense that the answer depends on the order of the coefficient.

For the time being, I'll leave it as it is, though it is a fair point to check those cases in detail because it may yield another performance improvement.

File: src/auxiliary.jl (hunk 11 above)
Comment (lbenet) [inline src/auxiliary.jl]: I don't think we require checking the bounds because, if we are out of the bounds, an error will be thrown.

File: src/auxiliary.jl (hunk 12 above)
Comment (lbenet) [inline src/auxiliary.jl]: I included the `copy` (knowing about the penalty that this entails) because without it you can change the coefficients of `a` (or `b`) and affect the *original* `a` and `b`, i.e., produce a side effect. 

Probably if `a.order == b.order` should also `return $T(copy(a.coeffs), a.order), $T(copy(b.coeffs), b.order)`, but for the moment, that is where `fixorder` is used, there are no side effects. Should I write that case with the `copy`s, since we are filtering the case `a.order == b.order` ? (You commented on this previously.)

File: src/constructors.jl (hunk 14 above)
Comment (lbenet) [inline src/constructors.jl]: Since there are no warnings for both Julia 0.5 and 0.6, I guess not...

File: src/power.jl (hunk 15 above)
Comment (lbenet) [inline src/power.jl]: Previously, in  487cb8b, we had this function, which turns out that it improves the performance with respect to using `Base.power_by_squaring`. 

The reason is that, wherever you read `square(x)`, in Base it is written as `x*x`, and `square(x)` is faster and uses less memory.

File: src/power.jl (hunk 16 above)
Comment (lbenet) [inline src/power.jl]: It could *almost* be replaced by `mul!(c,a,a,kend)`, but it can't. The reason is that the core of the loop is identical to the loop in `mul!`, but it updates `c[k+1]` instead of `c[kend+1]` as `mul!` would do.

File: src/power.jl (hunk 17 above)
Comment (lbenet) [inline src/power.jl]: `sqr!` for scalars will be non-mutating, so you will have to return the updated value. Then, either you keep the `if`, or return a corresponding value for `sqr!` when it acts on `HomogenousPolynomials`; the latter will have a cost in memory...

File: test/fateman40.jl (hunk 18 above)
Comment (lbenet) [inline test/fateman40.jl]: Same reason as above; get rid of a method that had side effects.

**Comment by lbenet** [comment]
@dpsanders Thanks a lot for the review; I just finished answering all your comments. As a result of your review, I pushed another commit that inlines `fixorder`. My local tests show that this improves further some benchmarks.

I'll squash it and merge it...
